File: mlp-ga/scripts/Neuron.js
class Neuron {

    constructor(numberOfConnections, weights, bias) {
        if((weights === null) || (bias === null)) {
            this.weights = [];
            for(var i = 0; i < numberOfConnections; i++) {
                this.weights.push(Neuron.randomWeight());
            }
            this.bias = Math.random();
            this.output = 0;
        } else {
            this.weights = weights;
            this.bias = bias;
            this.output = 0;
        }
    }

    static randomWeight() {
        return 2 * Math.random() - 1;;
    }

    calculateOutput(inputs) {
        var weightedInput = this.calculateWeightedInput(inputs);
        weightedInput += this.bias;
        this.output = 1 / (1 + Math.exp(-weightedInput));
    }

    calculateWeightedInput(inputs) {
        var ret = 0;
        for(var i = 0; i < inputs.length; i++) {
            ret += inputs[i] * this.weights[i];
        }
        return ret;
    }

    getOutput() {
        return this.output;
    }

    print() {
        console.log("weights = " + this.weights);
        console.log("output = " + this.output);
        console.log("bias = " + this.bias);
    }

    static reproduce(neuron1, neuron2, mutationRate) {
        var weights = [];
        for(var i = 0; i < neuron1.weights.length; i++) {
            if(Math.random() < 0.5) {
                weights.push(neuron1.weights[i]);
            } else {
                weights.push(neuron2.weights[i]);
            }
            weights[weights.length - 1] = Neuron.randomDeltaMutation(weights[weights.length - 1], mutationRate);
        }
        var bias;
        if(Math.random() < 0.5) {
            bias = neuron1.bias;
        } else {
            bias = neuron2.bias;
        }
        bias = Neuron.randomDeltaMutation(bias, mutationRate);
        return new Neuron(null, weights, bias);
    }

    static randomDeltaMutation(value, mutationRate) {
        var deltaValue;
        if(Math.random() < mutationRate) { 
            deltaValue = 2 * Math.random() - 1;
        } else {
            deltaValue = 0;
        }
        
        return value + deltaValue;
    }

}

File: java/main/neuron/src/main/java/com/yzumis/ia/neuron/Neuron.java
package com.yzumis.ia.neuron;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Yzumi on 27/12/2017.
 */
public class Neuron {

    private final List<Double> weights;
    private final double bias;
    private double output;

    public Neuron(final int numberOfconnections) {
        this.weights = new ArrayList<>();
        for(int i = 0; i < numberOfconnections; i++) {
            this.weights.add(this.generateRandomWeight());
        }
        this.bias = this.generateRandomWeight();
    }

    public Neuron(final List<Double> weights, final double bias) {
        this.weights = weights;
        this.bias = bias;
    }

    private double generateRandomWeight() {
        return Math.random();
    }

    public void calculateOutput(final List<Double> inputs) {
        final double weightedInput = this.calculateWeightedInput(inputs) + this.bias;
        this.output = 1.0 / (1.0 + Math.exp(-weightedInput));
    }

    private double calculateWeightedInput(final List<Double> inputs) {
        double ret = 0;
        for(int i = 0; i < inputs.size(); i++) {
            ret += this.weights.get(i) * inputs.get(i);
        }
        return ret;
    }

    public double getOutput() {
        return output;
    }

    public static Neuron reproduce(final Neuron neuron1, final Neuron neuron2, final double mutationRate) {
        final List<Double> childWeights = new ArrayList<>();
        for(int i = 0; i < neuron1.weights.size(); i++) {
            double childWeight;
            if(Math.random() < 0.5) {
                childWeight = neuron1.weights.get(i);
            } else {
                childWeight = neuron2.weights.get(i);
            }
            childWeight = Neuron.randomDeltaMutation(childWeight, mutationRate);
            childWeights.add(childWeight);
        }
        double bias;
        if(Math.random() < 0.5) {
            bias = neuron1.bias;
        } else {
            bias = neuron2.bias;
        }
        bias = Neuron.randomDeltaMutation(bias, mutationRate);
        return new Neuron(childWeights, bias);
    }

    private static double randomDeltaMutation(final double value, final double mutationRate) {
        final double deltaValue;
        if(Math.random() < mutationRate) {
            deltaValue = 2 * Math.random() - 1;
        } else {
            deltaValue = 0;
        }
        return value + deltaValue;
    }

}